linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4683,4 +4683,191 @@ private void flush() {
       }
     };
   }
+
+  /**
+   * Merge Union Enumerable.
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   *
+   * @param sources input enumerables (must be already sorted)
+   * @param sortKeySelector sort key selector
+   * @param sortComparator sort comparator to decide the next item
+   * @param all whether duplicates will be considered or not
+   * @param equalityComparer {@link EqualityComparer} to control duplicates,
+   *                         only used if {@code all} is {@code false}
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  public static <TSource, @Nullable TKey> Enumerable<TSource> mergeUnion(
```

Comment:
Did you really mean `TKey must always be nullable`?

Just in case:

`public static <@Nullable TKey>` means `all instantiations must use nullable type for the key`.

`public static <TKey extends @Nullable Object>` means `users could use either nullable or non-nullable implementations`

`public static <TKey>` is the same as `public static <TKey extends @Nullable Object>`.

`public static <TKey extends Object>`  means `TKey must be non-nullable`.

Reply:
If I don't annotate the TKey I get the following error:
````
error: [argument.type.incompatible] incompatible argument for parameter arg0 of compare.
      return this.sortComparator.compare(key1, key2);
                                         ^
  found   : TKey[ extends @Initialized @Nullable Object super @Initialized @Nullable Void]
  required: TKey[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
```
How should I annotate the TKey?

Reply:
```java
    private int compare(TSource e1, TSource e2) {
      final TKey key1 = e1 == null ? null : this.sortKeySelector.apply(e1);
      final TKey key2 = e2 == null ? null : this.sortKeySelector.apply(e2);
      return this.sortComparator.compare(key1, key2);
    }
```

Ok, what does the error mean here?
It says that the comparison key can become `null` even in the case `sortKeySelector` never returns nulls. I believe if the user passes `sortKeySelector` which never returns nulls, they won't expect that `mergeUnion` would silently skip `keySelector`.

Unfortunately, you do not declare the behavior with regards to the `null` values for `TSource` elements in `List<Enumerable<TSource>> sources`.

---

I believe the way to fix this is to remove custom `null` handling and just pass the keys to `sortKeySelector` as is:

```java
    private int compare(TSource e1, TSource e2) {
      final TKey key1 = this.sortKeySelector.apply(e1);
      final TKey key2 = this.sortKeySelector.apply(e2);
      return this.sortComparator.compare(key1, key2);
    }
```

Reply:
I fear that you suggestion might trigger a NPE inside `sortKeySelector.apply` if we pass a null element e1/e2.
BTW `sortComparator.compare` should be able to handle (in some cases) null keys (e.g. when we have nulls first or nulls last, and the comparator must handle nulls accordingly when compared to a non-null value).

Reply:
Currently you declare `sortKeySelector` as `Function1<TSource, TKey> sortKeySelector`.
In other words, it **must** work for all `TSource` values no matter if `TSource` is null or not.

On the other hand, if you implement custom `null` behavior, you basically augment user-provided `sortKeySelector`, and you basically forbid users to handle `nulls` differently.

I believe this silent implementation detail is worse than a clear NPE in case the user-provided function fails to handle nulls.

Reply:
I understand. The complication is that this `sortKeySelector` is / can be created by dynamic code (`PhysTpe#generateCollationKey`), and (it's not so easy to guess) I believe a null in there can lead to a NPE.
I have committed a new workaround based on `@SuppressWarnings("argument.type.incompatible")`

Reply:
> and (it's not so easy to guess) I believe a null in there can lead to a NPE.

`linq4j` is a general-purpose library, and it must not depend on `PhysType` quirks

Reply:
The problem with `e1 == null ? null` is that behavior is not documented, and it is not tested (there's no test which checks what happens in case user-provided `keySelector` wants to convert `null` to something non-`null`).

`mergeUnion` method is already very generic (it has customization for `keySelector`, `comparator`), so I see no reasons why the method should have its own `null-handling`. It is generic enough to let users decide what they want to do with nulls.

Reply:
You're right, in other parts of `EnumerableDefaults` using `Function1<TSource, TKey> keySelector` there is no null check before calling `keySelector.apply`, I'll remove the null handling here.

Reply:
Change committed. Thanks for your input @vlsi